pkg/endpoint/public_ip_watcher.go

```diff
@@ -65,26 +66,39 @@ func (p *PublicIPWatcher) Run(stopCh <-chan struct{}) {
 func (p *PublicIPWatcher) syncPublicIP() {
 	localEndpointSpec := p.config.LocalEndpoint.Spec()
 
-	publicIP, resolver, err := getPublicIP(p.config.SubmSpec, p.config.K8sClient, localEndpointSpec.BackendConfig, false)
+	publicIPs, err := getPublicIPs(p.config.SubmSpec, p.config.K8sClient, localEndpointSpec.BackendConfig, false, false)
 	if err != nil {
-		logger.Warningf("Could not determine public IP of the gateway node %q: %v", localEndpointSpec.Hostname, err)
+		logger.Warningf("Could not determine public IPs of the gateway node %q: %v", localEndpointSpec.Hostname, err)
 		return
 	}
 
-	if localEndpointSpec.PublicIP != publicIP {
-		logger.Infof("Public IP changed for the Gateway, updating the local endpoint with public IP %q obtained from resolver %q",
-			publicIP, resolver)
+	publicIpsChanged := false
 
-		if err := p.updateLocalEndpoint(publicIP); err != nil {
+	for _, publicIP := range publicIPs {
+		if (k8snet.IPFamilyOfString(publicIP) == k8snet.IPv4) && (publicIP != localEndpointSpec.GetPublicIP(k8snet.IPv4)) {
+			publicIpsChanged = true
+		}
+
+		if (k8snet.IPFamilyOfString(publicIP) == k8snet.IPv6) && (publicIP != localEndpointSpec.GetPublicIP(k8snet.IPv6)) {
+			publicIpsChanged = true
+		}
```

Comment:
I think this code is simplified if it loops over the families and passes it to `getPublicIP` as I suggested in my code snippet.  This also avoids the `logPublicIP` flag which seems awkward.